docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+like this:
+
+- All Tenant objects hold a buffer of object keys to delete, which is flushed to the next step whenever
+  they write their index file (i.e. step 2 in the steps above)
+- On flush of the per-tenant buffer, keys to delete are gathered into a shared DeletionAccumulator,
+  which writes out deletion lists to S3 in batches of some size chosen for efficiency.
+- A background task wakes up and processes deletion lists. Before executing the actual deletion, it must call
+  into the control plane (step 3 in the steps above).
+
+Deletion lists in S3 may _not_ be executed as-is by future writers for the tenant: they
+are only valid within a particular node generation. Future writers may either drop the deletion
+list and let orphan object cleanup handle the objects (see below), or they may reconcile the
+deletion objects with latest metadata to ensure all victim objects are really not referenced by
+latest metadata.
+
+### Housekeeping: Cleaning up orphan objects
+
+An orphan object is any object which is no longer referenced by a running node or by metadata.
+
+Examples of how orphan objects arise:
+
+- A node is doing compaction and writes a layer object, then crashes before it writes the
+  index_part.json that references that layer.
+- A partition node carries on running for some time, and writes out an unbounded number of
+  objects while it believes itself to be the rightful writer for a tenant.
+
+Orphan objects are functionally harmless, but have a small cost due to S3 capacity consumed. We
+may clean them up at some time in the future, but doing a ListObjectsv2 operation and cross
+referencing with the
+
+### Index changes
+
+IndexPart currently stores keys and LSNs sufficient to reconstruct key names: this would be
+extended to store the generation numbers with which a particular object was written, to enable
+reconstructing the suffix as well.
+
+This will increase the size of the file, but only modestly: layers are already encoded as
+their string-ized form, so the overhead is about 20 bytes per layer. This will be less if/when
+the index storage format is migrated to a binary format from JSON.
+
+### Pageserver startup changes
+
+- The pageserver must obtain a generation number by some means (see Control Plane Changes below) before
+  doing any remote writes.
+- For efficiency, the pageserver _should_ check that its attachments are still valid before proceeding
+  to do any writes. Writes to stale attachments are safe, but wasteful. If all the attachments are
+  the same as before restart, then it is not necessary to re-attach anything: the pageserver may proceed
+  with the same attachment generation as before (safe because the node generation has incremented)
+
+### `pageserver::Tenant` startup changes
+
+Within the pageserver, a tenant's startup must account for the possibility that this node was
+not the last node to write to the tenant's remote storage, and therefore execute a recovery to
+get an up to date LayerMap.
```

Comment:
What will go wrong if we skip that step?

Reply:
Added an explanation inline:
> This is necessary because if another node updated
the remote storage, it might have advanced the remote_consistent_lsn such that the safekeeper drops WAL
content that would be needed for this node to recover. It also acts as an optimization, to avoid redundantly
re-ingesting WAL data that is already in S3.

Reply:
896bfdb2299d2a4f4b1b942a35f5704ce801d020

Reply:
896bfdb2299d2a4f4b1b942a35f5704ce801d020